execution-env.md

```diff
@@ -1,80 +1,120 @@
-1.  The App host code invokes WebGPU APIs to create resources, load shaders, and create a pipeline configuration.
-1.  The App invokes a WebGPU API to run a pipeline.  In response, the WebGPU implementation:
-    1.  Validates the pipeline configuration and all associated objects. If the configuration is invalid, this procedure terminates with an error.  Otherwise the procedure continues.
-    1.  Determines the set of data resources that should be accessible to the shaders, and in what access modes (e.g. read-only, read-write).
-    1.  Determines the computing resources to be used to execute the pipeline.
-    1.  Creates a pipeline ExecutionContext consisting of the computing and data resources that are valid to be used by the pipeline, and in particular by the shaders in the pipeline.  Each data resource is associated with valid access modes: read-only resources may only be read; only writable resources may be written.
-    1.  Executes the pipeline, subject to the security property that follows.
+1. Validated the pipeline configuration and all associated objects.
+1. Determined the set of data resources that should be accessible to the shaders, and in what access modes (e.g. read-only, read-write).
+1. Determined the computing resources to be used to execute the pipeline.
+1. Created a pipeline ExecutionContext consisting of the computing and data resources that are valid to be used by the pipeline, and in particular by the shaders in the pipeline. 
+   Each data resource is associated with valid access modes: read-only resources may only be read; only writable resources may be written.
+1. Executes the pipeline, subject to the security property that follows.
 
 A WebGPU implementation is _secure_ if, for any application _App_ running on the implementation, accesses explicitly granted to App are the only means by which:
 
-
-
 *   _App_ may sense state outside its ExecutionContext
 *   _App_ may cause mutation of state outside its ExecutionContext
 *   _App_ may cause or sense I/O outside its ExecutionContext
 
 An access is considered explicitly granted only if:
 
-
-
 *   It is a read access to a resource attached for reading to the ExecutionContext
 *   It is a write access to a resource attached for writing to the ExecutionContext
 
+Typically a WebGPU implementation can make sure that the _App_ can only access resources provided in the ExecutionContext by validating the code in the shader module.
+However this only guarantees that the correct resources are accessed, not that they are accessed in bounds.
+This is an issue because resource accesses are often lowered to pointer arithmetic so out of bounds accesses could result in accesses outside the ExecutionContext.
+For this reason out of bound accesses to resources must be prevented.
 
-## WebGPU Execution Environment Specification for SPIR-V
+Validating accesses are in bounds at shader module or pipeline creation time isn't possible because it is equivalent to the halting problem.
```

Comment:
I wouldn't say this.  I get what you're trying to say, and maybe this is true with some caveats, but it would be easy for someone to take offense. For example, some Ada dialect has statically verified bounds on all accesses, which they achieve through language restrictions that wouldn't be practical for us. However, I don't think that this Ada dialect solves the halting problem.

I would say: "Validating accesses are in bounds at shader module or pipeline creation time is not possible for all valid SPIR-V programs."

Reply:
How about "Validating accesses are in bounds at shader module or pipeline creation time is not possible for all valid shader programs without making the programming model too inflexible."? It encodes your comment about the Ada dialect, and explains this isn't specific to SPIR-V.